include/laser_filters/scan_blob_filter.h

```diff
@@ -0,0 +1,158 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+/*
+  \author Kei OKada
+
+
+*/
+
+#ifndef LASER_SCAN_BLOB_FILTER_H
+#define LASER_SCAN_BLOB_FILTER_H
+
+#include <set>
+
+#include "filters/filter_base.h"
+#include <sensor_msgs/LaserScan.h>
+#include "angles/angles.h"
+
+namespace laser_filters{
+
+/** @b ScanBlobFilter is a simple filter that filters shadow points in a laser scan line 
+ */
+
+class ScanBlobFilter : public filters::FilterBase<sensor_msgs::LaserScan>
+{
+public:
+
+  double max_radius_;          // Filter angle threshold
+  int min_points_;
+    
+
+  ////////////////////////////////////////////////////////////////////////////////
+  ScanBlobFilter () 
+  {
+
+
+  }
+
+  /**@b Configure the filter from XML */
+  bool configure()
+  {
+    max_radius_ = 0.1;//default value
+    if (!filters::FilterBase<sensor_msgs::LaserScan>::getParam(std::string("max_radius"), max_radius_))
+    {
+      ROS_ERROR("Error: BlobFilter was not given min_radius.\n");
+      return false;
+    }
+
+    min_points_ = 5;//default value
+    if (!filters::FilterBase<sensor_msgs::LaserScan>::getParam(std::string("min_points"), min_points_))
+    {
+      ROS_INFO("Error: BlobFilter was not given min_points.\n");
+      return false;
+    }
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  virtual ~ScanBlobFilter () { }
+
+  ////////////////////////////////////////////////////////////////////////////////
+  /** \brief 
+   * \param scan_in the input LaserScan message
+   * \param scan_out the output LaserScan message
+   */
+  bool update(const sensor_msgs::LaserScan& scan_in, sensor_msgs::LaserScan& scan_out)
+  {
+    //copy across all data first
+    scan_out = scan_in;
+
+    std::set<int> indices_to_publish;
+    // assume that all points is pass thorugh shadow filter, so each blob is separeted by invalide scan data
+    std::vector<std::vector<int> > range_blobs;
+    std::vector<int> range_blob;
+    for (unsigned int i = 0; i < scan_in.ranges.size (); i++)
+    {
+      scan_out.ranges[i] = -1.0 * fabs(scan_in.ranges[i]); // set all ranges to invalid (*)
+      if ( scan_in.ranges[i] < 0 || std::isnan(scan_in.ranges[i])) {
+          if ( range_blob.size() > min_points_ ) {
+              range_blobs.push_back(range_blob);
+          }
+          range_blob.clear();
+      }else{
+          range_blob.push_back(i);
+      }
+    }
+    if ( range_blob.size() > min_points_ ) {
+        range_blobs.push_back(range_blob);
+    }
+    // for each blob calculate center and radius
+    for (unsigned int i = 0; i < range_blobs.size(); i++) {
+        int size = range_blobs[i].size();
+        // check center of blob
+        double center_x = 0, center_y = 0;
+        for (unsigned int j = 0; j < size; j++) {
+            double x = scan_in.ranges[range_blobs[i][j]];
+            double y = scan_in.ranges[range_blobs[i][j]] * scan_in.angle_increment;
```

Comment:
scan_shadows filter is just passing that through to `getAngleWithViewpoint()`, which i guess expects that. But using it here as center_y doesn't make any sense to me.... what do you want `y` to be? If it isn't going to be the cartesian `y` coordinate, you need to document what it actually is.